mirrord/agent/src/metrics.rs

```diff
@@ -0,0 +1,177 @@
+    register_int_gauge!(
+        "mirrord_agent_mirror_connection_subscription_count",
+        "amount of connections in mirror mode in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_FILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_steal_filtered_port_subscription_count",
+        "amount of filtered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_UNFILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_steal_unfiltered_port_subscription_count",
+        "amount of unfiltered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_FILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of filtered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static STEAL_UNFILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of unfiltered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static TCP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_tcp_outgoing_connection_count",
+        "amount of tcp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static UDP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_udp_outgoing_connection_count",
+        "amount of udp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+#[derive(Error, Debug)]
+pub(crate) enum MetricsError {
+    #[error(transparent)]
+    FromUtf8(#[from] std::string::FromUtf8Error),
+
+    #[error(transparent)]
+    Prometheus(#[from] prometheus::Error),
+}
+
+impl IntoResponse for MetricsError {
+    fn into_response(self) -> axum::response::Response {
+        (http::StatusCode::INTERNAL_SERVER_ERROR, self.to_string()).into_response()
+    }
+}
+
+#[tracing::instrument(level = Level::TRACE,  ret, err)]
+async fn get_metrics() -> Result<String, MetricsError> {
+    use prometheus::{Encoder, TextEncoder};
+
+    let metric_families = prometheus::gather();
+
+    let mut buffer = Vec::new();
+    TextEncoder
+        .encode(&metric_families, &mut buffer)
+        .inspect_err(|error| tracing::error!(%error, "unable to encode prometheus metrics"))?;
+
+    Ok(String::from_utf8(buffer)?)
```

Comment:
You can use [`TextEncode::encode_to_string`](https://docs.rs/prometheus/latest/prometheus/struct.TextEncoder.html#method.encode_to_string) (you won't need `MetricsError` anymore)

Reply:
It also returns a `prometheus::error`, how can I avoid `MetricsError` with it?

Reply:
I'd just change the return type to `(StatusCode, String)`. The function could be very small, just
```rust
let metric_families = prometheus::gather();
match TextEncoder.encode_to_string(&metric_families) {
    Ok(response) => (StatusCode::OK, response),
    Err(error) => {
        tracing::error!(...);
        (StatusCode::INTERNAL_SERVER_ERROR, error.to_string())
    }
}
```